ballerina/executor_visitor.bal

```diff
@@ -15,76 +15,132 @@
+                    self.data[fieldNode.getAlias()] = MUTATION_TYPE_NAME;
+                } else {
+                    self.data[fieldNode.getAlias()] = SUBSCRIPTION_TYPE_NAME;
+                }
             } else {
-                self.data[fieldNode.getAlias()] = SUBSCRIPTION_TYPE_NAME;
+                isIntrospection = false;
             }
-        } else {
-            self.execute(fieldNode, operationType, self.result);
+        }
+        function (parser:FieldNode, parser:RootOperationType) execute;
+        lock {
+            execute = self.execute;
+        }
+        if !isIntrospection {
+            execute(fieldNode, operationType);
         }
     }
 
     public isolated function visitArgument(parser:ArgumentNode argumentNode, anydata data = ()) {}
 
     public isolated function visitFragment(parser:FragmentNode fragmentNode, anydata data = ()) {
+        parser:RootOperationType operationType = <parser:RootOperationType>data;
+        if operationType != parser:OPERATION_MUTATION {
+            return self.visitSelectionsParallelly(fragmentNode, operationType);
+        }
         foreach parser:SelectionNode selection in fragmentNode.getSelections() {
-            selection.accept(self, data);
+            selection.accept(self, operationType);
+        }
+    }
+
+    private isolated function visitSelectionsParallelly(parser:SelectionParentNode selectionParentNode,
+                                                       anydata data = ()) {
+        future<()>[] futures = [];
+        readonly & anydata clonedData = data.cloneReadOnly();
+        foreach parser:SelectionNode selection in selectionParentNode.getSelections() {
+            future<()> 'future = start selection.accept(self, clonedData);
+            futures.push('future);
+        }
+        foreach future<()> 'future in futures {
+            _ = checkpanic wait 'future;
         }
     }
 
     public isolated function visitDirective(parser:DirectiveNode directiveNode, anydata data = ()) {}
 
     public isolated function visitVariable(parser:VariableNode variableNode, anydata data = ()) {}
 
-    isolated function execute(parser:FieldNode fieldNode, parser:RootOperationType operationType, any|error fieldValue) {
-        Field 'field = getFieldObject(fieldNode, operationType, self.schema, self.engine, fieldValue = fieldValue);
-        self.context.resetInterceptorCount();
-        var result = self.engine.resolve(self.context, 'field);
-        self.errors = self.context.getErrors();
-        self.data[fieldNode.getAlias()] = result is ErrorDetail ? () : result;
+    isolated function execute(parser:FieldNode fieldNode, parser:RootOperationType operationType) {
+        any|error result;
+        __Schema schema;
+        Engine engine;
+        Context context;
+        lock {
+            result = self.getResult();
+            schema = self.schema.clone();
+            engine = self.engine;
+            context = self.context;
+        }
```

Comment:
Why do we need to clone these again? We are cloning this in the init method, right? Can't we use those values (by cloning them readonly)?

Reply:
+1, removed made schema readonly